### PR TITLE
[ROCm] Switch on Triton feature for ROCm.

### PR DESCRIPTION
**Pull request by zoranjovanovic-ns**

### CODE REVIEW

**Review by sergeykozub — APPROVED** [review]

**Comment by zoranjovanovic-ns** [comment]
Hi @xla-rotation, CI is green, Do I need something more to have this merged?

File: xla/service/gpu/BUILD [hunk 1]
```diff
@@ -588,6 +588,9 @@ cc_library(
         "@triton//:TritonLLVMIR",
     ]) + if_rocm_is_configured([
         "@tsl//tsl/platform:rocm_rocdl_path",
+        "@triton//:TritonLLVMIR",
```
Comment (chsigg) [inline xla/service/gpu/BUILD:591]: This should move to `if_gpu_is_configured`.

File: third_party/triton/temporary/amd_pr7.patch [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+         "@llvm-project//mlir:Pass",
+         "@llvm-project//mlir:Support",
+-        "@llvm-project//mlir:TransformUtils",
++        "@llvm-project//mlir:ROCDLDialect",
+         "@llvm-project//mlir:Transforms",
+     ],
+ )
+diff --git a/third_party/amd/lib/TritonAMDGPUToLLVM/ElementwiseOpToLLVM.cpp b/third_party/amd/lib/TritonAMDGPUToLLVM/ElementwiseOpToLLVM.cpp
+index f59efd6..cf601f0 100644
+--- a/third_party/amd/lib/TritonAMDGPUToLLVM/ElementwiseOpToLLVM.cpp
```
Comment (chsigg) [inline third_party/triton/temporary/amd_pr7.patch:23]: Are there any plans to upstream this? I couldn't find the PR.

File: xla/service/gpu/gemm_fusion.cc [hunk 3]
```diff
@@ -801,10 +801,13 @@ absl::StatusOr<bool> GemmFusion::Run(
...
       std::get_if<se::CudaComputeCapability>(&gpu_version_);
-  if (!cuda_compute_capability) {
+  auto rocm_compute_capability =
+      std::get_if<se::RocmComputeCapability>(&gpu_version_);
+  if (!cuda_compute_capability && !rocm_compute_capability) {
     return absl::FailedPreconditionError(
-        "Triton support is only enabled for CUDA GPUs.");
-  } else if (!cuda_compute_capability->IsAtLeastAmpere()) {
+        "Triton support is only enabled for CUDA and ROCm GPUs.");
+  } else if (cuda_compute_capability != nullptr &&
```
Comment (chsigg) [inline xla/service/gpu/gemm_fusion.cc:809]: `cuda_compute_capability` instead of `cuda_compute_capability != nullptr`.

Comment (chsigg) [inline xla/service/gpu/gemm_fusion.cc:809]: Please remove `else` after return.

File: xla/service/gpu/gpu_compiler.cc [hunk 4]
```diff
@@ -1379,16 +1379,21 @@ absl::Status GpuCompiler::OptimizeHloPostLayoutAssignment(
...
                                    /*f8_rewrite=*/true);
-    if (debug_options.xla_gpu_enable_triton_gemm() && cuda_cc != nullptr &&
-        cuda_cc->IsAtLeast(se::CudaComputeCapability::AMPERE)) {
+    if (debug_options.xla_gpu_enable_triton_gemm() && ((cuda_cc != nullptr &&
+        cuda_cc->IsAtLeast(se::CudaComputeCapability::AMPERE)) ||
+        rocm_cc != nullptr)) {
       pipeline.AddPass<GemvRewriter>();
       pipeline.AddPass<GemmFusion>(gpu_version);
     }
+    if (debug_options.xla_gpu_enable_triton_gemm() && rocm_cc != nullptr) {
```
Comment (chsigg) [inline xla/service/gpu/gpu_compiler.cc:1394]: This looks fishy, do you really want to add the same pass twice?

File: third_party/triton/temporary/amd_pr7.patch (hunk 2 above)
Comment (zoranjovanovic-ns) [reply third_party/triton/temporary/amd_pr7.patch:23]: Yes, I will prepare PR for build system changes.

**Comment by zoranjovanovic-ns** [comment]
@chsigg Applied comments and updated PR.

**Comment by penpornk** [comment]
@zoranjovanovic-ns This branch has conflicts. Could you please help resolve them as well?

**Comment by zoranjovanovic-ns** [comment]
> @zoranjovanovic-ns This branch has conflicts. Could you please help resolve them as well?

Yes, I am working on it.

**Review by chsigg — APPROVED** [review]

**Pull request closed without merging**